partitura/performance.py

```diff
@@ -280,6 +284,127 @@ def adjust_offsets_w_sustain(
+        If not provided, the default values will be used.
+        Pitch, note_on, and note_off are required.
+    """
+
+    def __init__(self, pnote_dict):
+        super().__init__(pnote_dict)
+        self["id"] = self.get("id", None)
+        self["pitch"] = self.get("pitch", self["midi_pitch"])
+        self["note_on"] = self.get("note_on", -1)
+        self["note_off"] = self.get("note_off", -1)
+        self["sound_off"] = self.get("sound_off", self["note_off"])
+        self["track"] = self.get("track", 0)
+        self["channel"] = self.get("channel", 1)
+        self["velocity"] = self.get("velocity", 60)
+        self._validate_values()
+        self._accepted_keys = [
+            "id",
+            "pitch",
+            "note_on",
+            "note_off",
+            "velocity",
+            "track",
+            "channel",
+            "sound_off",
+        ]
+        self.__setitem__ = self._setitem_new
+
+    def __str__(self):
+        return f"PerformedNote: {self['id']}"
+
+    def __eq__(self, other):
+        if not isinstance(PerformedNote):
+            return False
+        if not self.keys() == other.keys():
+            return False
+        return np.all(
+            np.array([self[k] == other[k] for k in self.keys() if k in other.keys()])
+        )
+
+    def __hash__(self):
+        return hash(self["id"])
+
+    def __lt__(self, other):
+        return self["note_on"] < other["note_on"]
+
+    def __le__(self, other):
+        return self["note_on"] <= other["note_on"]
+
+    def __gt__(self, other):
+        return self["note_on"] > other["note_on"]
+
+    def __ge__(self, other):
+        return self["note_on"] >= other["note_on"]
+
+    def __getitem__(self, key):
+        return self.get(key, None)
+
+    def _setitem_new(self, key, value):
+        if key not in self._accepted_keys:
+            raise KeyError(f"Key {key} not in PerformedNote")
+        elif key == "note_off":
+            # Verify that the note_off is after the note_on
+            if value < self["note_on"]:
+                raise ValueError(f"note_off must be after or equal to note_on")
+            self["sound_off"] = (
+                value if self["sound_off"] < value else self["sound_off"]
+            )
+            self["note_off"] = value
+        elif key == "note_on":
+            # Verify that the note_on is before the note_off
+            if value > self["note_off"]:
+                raise ValueError(f"note_on must be before or equal to note_off")
+
+            self["duration_sec"] = self["note_off"] - value
+            self["note_on"] = value
+        elif key == "sound_off":
+            # Verify that the sound_off is after the note_off
+            if value <= self["note_off"]:
+                raise ValueError(f"sound_off must be after or equal to note_off")
+            self["sound_off"] = value
+        else:
+            self[key] = value
```

Comment:
this function is never reached in my tests. if I rename is directly to `__setitem__` (is there a reason to use the standard setter in the init?), it breaks because it's full of recursions, e.g., line 378 `self[key] = value` calls `self.__setitem__` which should point to itself.

Reply:
Yes, you are right. I wrote this but it doesn't make a lot of sense. What if we overwrite the `__setitem__` to include a `validate_values` with a key (to not check everything all the time) if it is provided:
```python
def __setitem__(self, key, value):
   self[key] = value
   self.validate_values(key)
```